### PR TITLE
Equality visitor

### PR DESCRIPTION
**Pull request by a-zakir**
draft

### CODE REVIEW

File: src/solver/expressions/include/antares/solver/expressions/nodes/TwoNodesVisitor.h [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+template<class R>
+class TwoNodesVisitor
+{
+public:
+    virtual ~TwoNodesVisitor() = default;
+
+    R dispatch(const Node& node1, const Node& node2)
+    {
+        using FunctionT = std::optional<R> (*)(const Node&, const Node&, TwoNodesVisitor<R>&);
```
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/nodes/TwoNodesVisitor.h:56]: What about different node types ? Is there a default case for this ? For example `AddNode` and `MultiplicationNode`.

**Review by flomnes — COMMENTED** [review]
The key is to have something like
```cpp
bool operator==(const Node& node1, const Node& node2)
```
that works for all node types. I'm not sure your implementation does that.

File: src/tests/src/solver/expressions/test_expressions.cpp [hunk 2]
```diff
@@ -375,3 +376,37 @@ BOOST_FIXTURE_TEST_CASE(simple_constant_expression, Registry<Node>)
...
+    VariableNode y2("y");
+    // Mult : 56.*y
+    Node* v1 = create<MultiplicationNode>(&lit3, &y1);
+    Node* v2 = create<MultiplicationNode>(&lit4, &y2);
+
+    Node* add1 = create<AddNode>(u1, v1);
+    CloneVisitor cloneVisitor(*this);
+    Node* add1_clone = cloneVisitor.dispatch(*add1);
+
+    Node* add2 = create<AddNode>(u2, v2);
```
Comment (flomnes) [inline src/tests/src/solver/expressions/test_expressions.cpp]: Write a function to create the tree
```cpp
Node* root1 = createExpressionTree();
Node* root2 = createExpressionTree();
BOOST_CHECK(equalityVisitor.dispatch(*root1, *root2));
```

File: src/tests/src/solver/expressions/test_expressions.cpp [hunk 3]
```diff
@@ -375,3 +376,37 @@ BOOST_FIXTURE_TEST_CASE(simple_constant_expression, Registry<Node>)
...
+
+    PrintVisitor printVisitor;
+    EqualityVisitor equalityVisitor;
+
+    BOOST_CHECK_EQUAL(printVisitor.dispatch(*add1), printVisitor.dispatch(*add1_clone));
+    BOOST_CHECK_EQUAL(equalityVisitor.dispatch(*add1, *add2), true);
+    // TODO test depth
+    AddNode random_add(create<LiteralNode>(25.), create<VariableNode>("rand_var"));
```
Comment (flomnes) [inline src/tests/src/solver/expressions/test_expressions.cpp]: You may also compare a root to itself
```cpp
BOOST_CHECK(equalityVisitor.dispatch(*root, *root));
```

File: src/solver/expressions/include/antares/solver/expressions/nodes/TwoNodesVisitor.h [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+                                         &tryType<R, TwoNodesVisitor<R>, ComponentVariableNode>,
+                                         &tryType<R, TwoNodesVisitor<R>, ComponentParameterNode>};
+        for (auto f: functions)
+        {
+            if (auto ret = f(node1, node2, *this); ret.has_value())
+            {
+                return ret.value();
+            }
+        }
+        logs.error() << "Antares::Solver::Nodes Visitor: unsupported Node!";
```
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/nodes/TwoNodesVisitor.h]: If `node1` and `node2` are of different types, we will encounter this line. So I think we should handle errors better here.
Example
```cpp
Literal lit1(2), lit2(2);
AddNode node1(&lit1, &lit2);
MultiplicationNode node2(&lit1, &lit2);
EqualityVisitor equal;
// Here, we go through l. 78 above, but the caller only gets a log error
// and an `R()`, not so easy to catch
equal.dispatch(&node1, &node2); 
```

File: src/solver/expressions/include/antares/solver/expressions/nodes/TwoNodesVisitor.h [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+{
+    auto x = dynamic_cast<const NodeT*>(&node1);
+    auto y = dynamic_cast<const NodeT*>(&node2);
+    if (x && y)
+    {
+        return visitor.visit(*x, *y);
+    }
+    else
+    {
+        return {};
```
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/nodes/TwoNodesVisitor.h:43]: Different node types

File: src/solver/expressions/include/antares/solver/expressions/nodes/TwoNodesVisitor.h (hunk 1 above)
Comment (a-zakir) [reply src/solver/expressions/include/antares/solver/expressions/nodes/TwoNodesVisitor.h:56]: yes i have to check node consistency

File: src/tests/src/solver/expressions/test_expressions.cpp (hunk 2 above)
Comment (flomnes) [reply src/tests/src/solver/expressions/test_expressions.cpp]: BTW it's more like
```cpp
Node* root1 = createExpressionTree(registry);
```

File: src/solver/expressions/include/antares/solver/expressions/nodes/TwoNodesVisitor.h (hunk 4 above)
Comment (a-zakir) [reply src/solver/expressions/include/antares/solver/expressions/nodes/TwoNodesVisitor.h]: update: the code print warning message and returns false

File: src/solver/expressions/include/antares/solver/expressions/nodes/TwoNodesVisitor.h (hunk 5 above)
Comment (a-zakir) [reply src/solver/expressions/include/antares/solver/expressions/nodes/TwoNodesVisitor.h:43]: the return type is handled in the dispatch method

**Pull request closed without merging**